src/FESpaces/FESpacesWithDofPotentiallyRemoved.jl

```diff
@@ -0,0 +1,185 @@
+end
+
+function gather_free_and_dirichlet_values(
+  f::FESpaceWithDofPotentiallyRemoved{CS,true},cv) where {CS}
+  _fv, _dv = gather_free_and_dirichlet_values(f.space,cv)
+  @assert length(_dv) == 0
+  fv = vcat(view(_fv,1:f.dof_to_remove-1),
+            view(_fv,f.dof_to_remove+1:length(_fv))) # TODO: can we avoid
+                                          # allocating new memory?
+  dv = view(_fv,f.dof_to_remove:f.dof_to_remove)
+  (fv, dv)
+end
+
+function gather_free_and_dirichlet_values(
+  f::FESpaceWithDofPotentiallyRemoved{CS,false},cv) where {CS}
+  gather_free_and_dirichlet_values(f.space,cv)
+end
+
+function gather_free_and_dirichlet_values!(
+  fv,dv,f::FESpaceWithDofPotentiallyRemoved{CS,true},cv) where {CS}
+  _fv, _dv = gather_free_and_dirichlet_values(f.space,cv)
+  @assert length(_dv) == 0
+  fv .= vcat(view(_fv,1:f.dof_to_remove-1),
+            view(_fv,f.dof_to_remove+1:length(_fv))) # TODO: can we avoid
+                                          # allocating new memory?
+  dv .= view(_fv,f.dof_to_remove:f.dof_to_remove)
+  (fv, dv)
+end
+
+function gather_free_and_dirichlet_values!(
+  fv,dv,f::FESpaceWithDofPotentiallyRemoved{CS,false},cv) where {CS}
+  gather_free_and_dirichlet_values(f.space,cv)
+end
+
+function TrialFESpace(f::FESpaceWithDofPotentiallyRemoved{CS,RD}) where {CS,RD}
+  U = TrialFESpace(f.space)
+  FESpaceWithDofPotentiallyRemoved(U,RD,f.dof_to_remove)
+end
+
+# Delegated functions
+
+function get_cell_basis(f::FESpaceWithDofPotentiallyRemoved)
+  get_cell_basis(f.space)
+end
+
+function get_cell_dof_basis(f::FESpaceWithDofPotentiallyRemoved)
+  get_cell_dof_basis(f.space)
+end
+
+get_cell_axes(t::FESpaceWithDofPotentiallyRemoved)= get_cell_axes(t.space)
+
+get_cell_axes_with_constraints(t::FESpaceWithDofPotentiallyRemoved)= get_cell_axes_with_constraints(t.space)
+
+CellData.CellField(t::FESpaceWithDofPotentiallyRemoved,cell_vals) = CellField(t.space,cell_vals)
+
+constraint_style(::Type{<:FESpaceWithDofPotentiallyRemoved{CS}}) where CS = Val{CS}()
+
+# Helpers
+
+struct CellDofsWithDofRemoved{A<:AbstractArray} <: AbstractVector{Vector{Int}}
+  cell_dofs::A
+  dof_to_remove::Int
+end
+
+Base.size(a::CellDofsWithDofRemoved) = (length(a.cell_dofs),)
+
+Base.IndexStyle(::Type{<:CellDofsWithDofRemoved}) = IndexLinear()
+
+function Base.getindex(a::CellDofsWithDofRemoved,i::Integer)
+  cache = array_cache(a)
+  getindex!(cache,a,i)
+end
+
+function array_cache(a::CellDofsWithDofRemoved)
+  @assert eltype(a.cell_dofs) == Vector{Int}
+  b = testitem(a.cell_dofs)
+  c = CachedArray(b)
+  cache = array_cache(a.cell_dofs)
+  (c, cache)
+end
+
+@inline function getindex!(d,a::CellDofsWithDofRemoved,i::Integer)
+  c, cache = d
+  b = getindex!(cache,a.cell_dofs,i)
+  setsize!(c,size(b))
+  r = c.array
+  for j in 1:length(b)
+    bj = b[j]
+    if bj == a.dof_to_remove
+      rj = -1
+    elseif bj < a.dof_to_remove
+      rj = bj
+    else
+      rj = bj-1
+    end
+    r[j] = rj
+  end
+  r
+end
```

Comment:
Do we need this? Couldn't we just change the `data` vector in the `Table` that represents the face dofs using a lazy vector that only touches one entry?

Reply:
I think we need this (or something in this direction) to be general. We cannot assume in general that we have a `Table` for the face dofs. We could write (on top of this) a specialized version when we have a Table, but it seems premature optimization to me.

Reply:
It does not have to be a table but just create a method that modifies the `face_own_dofs` whatever it is. The idea is simple, only modify the plain global vector by touching one of, in the line of what you say above, e.g., `VectorWithIndexChanged` (horrible name) with the global Ids that is being used in `face_own_dofs`. This way, we do not need the code above, I guess.

Reply:
I worked on all the issues, except this one.

Just to be double-check, you propose to remove the `CellDofsWithDofFixed` data type, right?

Plus a possible alternative solution, that I do not enterily grasp. What I do not get is:
  *  "Probably create a method for adding these constraints in face_own_dofs Table."
  *  "It does not have to be a table but just create a method that modifies the face_own_dofs whatever it is." 

`face_own_dofs` is a Table returned by `ReferenceFE` right? How is the link among this Table and `get_cell_dofs()` such that we can achieve the desired effect?.  I am sure that they key is here, and thus I am missing something important.

I have been exploring `get_cell_dofs` references along the whole project, and there are a myriad of different data types that such method can return. Of course, all this data types can be conceptually seen as `Vector{Vector{Int}}`. Besides, such methods, can be traversed in several ways, using a single loop,  two nested loops, with and without caches, etc. 

What it comes into my mind is to define a new Vector of vectors, which is built from another vector of vectors, and two indices, `i`, `j`, and a value `v`, such that the new vector fulfills `x[i][j]==v`. But this is in some sense what `CellDofsWithDofFixed` is doing, so I expect to be missing something.